### PR TITLE
[tools] Don't require a LinkWith attributes to find Objective-C classes in assemblies.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Stop requiring a LinkWith attribute in an assembly in order to keep any Objective-C
types within. There are many ways to include a native library in a build nowadays,
and more and more often they don't need any LinkWith attributes to specify custom
linker behavior (in particular for frameworks, which can typically be included as-is).
The result of not searching such assemblies for Objective-C types would be that the
native linker would strip them away, and that would mean incorrect behavior at runtime.

However, this is a rather invasive change, especially for a minor release, so I'm
adding two things to make it better:

1. An opt-out MSBuild property: `RequireLinkWithAttributeForObjectiveCClassSearch`.
   Set to 'true' to opt-out (default is 'false').

2. Improve handling of native symbols with regards to the native linker.

    Add a new item group, ReferenceNativeSymbol, that contains native symbols
    we handle in some way - either to be ignored or we ask the native linker
    to keep it (by passing it as '-u ...' or in a symbol list file).

    There are two supported types of metadata:

    * SymbolType: either 'ObjectiveCClass', 'Function' or 'Field'. Used to
      compute the complete native name of a symbol (for instance, the native
      symbol for the Objective-C class "MyClass" is `_OBJC_CLASS_$_MyClass`,
      while for a function "MyFunction" it's just `_MyFunction`.
    * SymbolMode: either 'Ignore' or 'Default'. "Ignore" means to not pass the given
      symbol to the native linker, the default is to do so.

    SymbolType is required, while SymbolMode isn't.

    Example symbol to keep:

    ```xml
    <ItemGroup>
        <ReferenceNativeSymbol Include="MyClass" SymbolType="ObjectiveCClass" />
    </ItemGroup>
    ```

    Example symbol to ignore:

    ```xml
    <ItemGroup>
        <ReferenceNativeSymbol Include="MyClass" SymbolType="ObjectiveCClass" SymbolMode="Ignore" />
    </ItemGroup>
    ```

Finally use the latter solution to work around an issue that arouse with monotouch-test:
we reference an Objective-C class that doesn't exist in monotouch-test. This worked
because the referencing assembly didn't have a LinkWith attribute (and thus the reference
was ignored), but now that the reference isn't ignored anymore, we need to explicitly
ignore the Objective-C class.

### CODE REVIEW

File: tools/dotnet-linker/LinkerConfiguration.cs [hunk 1]
```diff
@@ -263,12 +263,47 @@ public static LinkerConfiguration GetInstance (LinkContext context)
...
 					PlatformAssembly = Path.GetFileNameWithoutExtension (value);
 					break;
+				case "ReferenceNativeSymbol": {
+					(string symbolType, string symbolMode, string symbol) = SplitString3 (value, ':');
+					var mode = SymbolMode.Default;
+					switch (symbolMode) {
+					case "Ignore":
+						mode = SymbolMode.Ignore;
+						break;
+					case "Default":
```
Comment (dalexsoto) [inline tools/dotnet-linker/LinkerConfiguration.cs:273]: Naming is hard! What is `Default`? :) maybe add a third value called `Include` and `Include` == `Default`? or just rename `Default` to `Include`? It is just not clear what `Default` means from just `Default` and `Ignore`

**Review by dalexsoto — COMMENTED** [review]
Looking good, please add a documentation file inside `docs`, this feature seems super useful for 3rd party bindings with missing symbols

File: tools/dotnet-linker/LinkerConfiguration.cs (hunk 1 above)
Comment (rolfbjarne) [reply tools/dotnet-linker/LinkerConfiguration.cs:273]: @dalexsoto I could just remove `"Default"` - then there would be two options: `"Ignore"` and `""` (empty string - which would really be "not ignore"). Maybe that makes more sense?

**Comment by rolfbjarne** [comment]
> Looking good, please add a documentation file inside `docs`, this feature seems super useful for 3rd party bindings with missing symbols

Documented here:

* https://github.com/xamarin/xamarin-macios/wiki/Project-file-properties#referencenativesymbol
* https://github.com/xamarin/xamarin-macios/wiki/Project-file-properties#requirelinkwithattributeforobjectivecclasssearch

File: tools/dotnet-linker/LinkerConfiguration.cs (hunk 1 above)
Comment (rolfbjarne) [reply tools/dotnet-linker/LinkerConfiguration.cs]: @dalexsoto I removed `Default` to simplify.

**Review by dalexsoto — APPROVED** [review]

**Pull request merged**